rust/agama-software/src/model/state.rs

```diff
@@ -0,0 +1,478 @@
+                    }
+                }
+            }
+        }
+
+        if let Some(only_required) = software.only_required {
+            state.options.only_required = only_required;
+        }
+    }
+
+    fn from_product_spec(&self) -> SoftwareState {
+        let software = &self.product.software;
+        let repositories = software
+            .repositories()
+            .into_iter()
+            .enumerate()
+            .map(|(i, r)| {
+                let alias = format!("agama-{}", i);
+                Repository {
+                    name: alias.clone(),
+                    alias,
+                    url: r.url.clone(),
+                    enabled: true,
+                }
+            })
+            .collect();
+
+        let mut patterns: Vec<Resolvable> = software
+            .mandatory_patterns
+            .iter()
+            .map(|p| Resolvable::new(p, false))
+            .collect();
+
+        patterns.extend(
+            software
+                .optional_patterns
+                .iter()
+                .map(|p| Resolvable::new(p, true)),
+        );
+
+        patterns.extend(software.user_patterns.iter().filter_map(|p| match p {
+            UserPattern::Plain(_) => None,
+            UserPattern::Preselected(pattern) => {
+                if pattern.selected {
+                    Some(Resolvable::new(&pattern.name, true))
+                } else {
+                    None
+                }
+            }
+        }));
+
+        SoftwareState {
+            product: software.base_product.clone(),
+            repositories,
+            patterns,
+            packages: vec![],
+            options: Default::default(),
+        }
+    }
+}
+
+impl SoftwareState {
+    // TODO: Add SoftwareSelection as additional argument.
+    pub fn build_from(product: &ProductSpec, config: &Config, system: &SystemInfo) -> Self {
+        SoftwareStateBuilder::for_product(product)
+            .with_config(config)
+            .with_system(system)
+            .build()
+    }
+}
+
+/// Defines a repository.
+#[derive(Debug)]
+pub struct Repository {
+    pub alias: String,
+    pub name: String,
+    pub url: String,
+    pub enabled: bool,
+}
+
+impl From<&RepositoryParams> for Repository {
+    fn from(value: &RepositoryParams) -> Self {
+        Repository {
+            name: value.name.as_ref().unwrap_or(&value.alias).clone(),
+            alias: value.alias.clone(),
+            url: value.url.clone(),
+            enabled: value.enabled.unwrap_or(true),
+        }
+    }
+}
+
+impl From<&packages::Repository> for Repository {
+    fn from(value: &packages::Repository) -> Self {
+        Repository {
+            name: value.name.clone(),
+            alias: value.alias.clone(),
+            url: value.url.clone(),
+            enabled: value.enabled,
+        }
+    }
+}
+
+/// Defines a resolvable to be selected.
+#[derive(Debug, PartialEq)]
+pub struct Resolvable {
```

Comment:
why not have a type here? It is kind of strange to have resolvable without type of it. It cannot be used standalone without context.